solidity/contracts/hooks/StaticProtocolFee.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+/*@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+     @@@@@  HYPERLANE  @@@@@@@
+    @@@@@@@@@@@@@@@@@@@@@@@@@
+   @@@@@@@@@       @@@@@@@@@
+  @@@@@@@@@       @@@@@@@@@
+ @@@@@@@@@       @@@@@@@@@
+@@@@@@@@@       @@@@@@@@*/
+
+// ============ Internal Imports ============
+import {Message} from "../libs/Message.sol";
+import {IPostDispatchHook} from "../interfaces/hooks/IPostDispatchHook.sol";
+// ============ External Imports ============
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+/**
+ * @title StaticProtocolFee
+ * @notice Collects a static protocol fee from the sender.
+ * @dev V3 WIP
+ */
+contract StaticProtocolFee is IPostDispatchHook, Ownable {
+    using Address for address payable;
+    using Message for bytes;
+
+    // ============ Constants ============
+
+    /// @notice The maximum protocol fee that can be set.
+    uint256 public immutable MAX_PROTOCOL_FEE;
+
+    // ============ Public Storage ============
+
+    /// @notice The current protocol fee.
+    uint256 public protocolFee;
+    /// @notice The beneficiary of protocol fees.
+    address public beneficiary;
+
+    // ============ Constructor ============
+
+    constructor(
+        uint256 _maxProtocolFee,
+        uint256 _protocolFee,
+        address _beneficiary,
+        address _owner
+    ) {
+        MAX_PROTOCOL_FEE = _maxProtocolFee;
+        _setProtocolFee(_protocolFee);
+        _setBeneficiary(_beneficiary);
+        _transferOwnership(_owner);
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Collects the protocol fee from the sender.
+     */
+    function postDispatch(bytes calldata, bytes calldata message)
+        external
+        payable
+        override
+    {
+        require(
+            msg.value >= protocolFee,
+            "StaticProtocolFee: insufficient protocol fee"
+        );
+
+        uint256 refund = msg.value - protocolFee;
+        if (refund > 0) payable(message.senderAddress()).sendValue(refund);
```

Comment:
wish there was an easy way to do this branchless
do we want to support a different refund address via metadata? similar to IGP?